src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -1115,57 +1068,35 @@ private async ValueTask WriteAsyncInternal<TIOAdapter>(TIOAdapter writeAdapter,
             }
         }
 
-        private void ConsumeBufferedBytes(int byteCount)
-        {
-            Debug.Assert(byteCount >= 0);
-            Debug.Assert(byteCount <= _internalBufferCount);
-
-            _internalOffset += byteCount;
-            _internalBufferCount -= byteCount;
-            if (_internalBufferCount == 0)
-            {
-                _internalOffset = 0;
-            }
-        }
-
         private int CopyDecryptedData(Memory<byte> buffer)
         {
-            Debug.Assert(_decryptedBytesCount > 0);
+            Debug.Assert(_buffer.DecryptedLength > 0);
 
-            int copyBytes = Math.Min(_decryptedBytesCount, buffer.Length);
+            int copyBytes = Math.Min(_buffer.DecryptedLength, buffer.Length);
             if (copyBytes != 0)
             {
-                new ReadOnlySpan<byte>(_internalBuffer, _decryptedBytesOffset, copyBytes).CopyTo(buffer.Span);
-
-                _decryptedBytesOffset += copyBytes;
-                _decryptedBytesCount -= copyBytes;
-            }
-
-            if (_decryptedBytesCount == 0)
-            {
-                _decryptedBytesOffset = 0;
+                _buffer.DecryptedReadOnlySpanSliced(copyBytes).CopyTo(buffer.Span);
+                _buffer.Discard(copyBytes);
             }
 
             return copyBytes;
         }
 
         private void ResetReadBuffer()
         {
-            Debug.Assert(_decryptedBytesCount == 0);
+           Debug.Assert(_buffer.DecryptedLength == 0);
 
-            if (_internalBuffer == null)
+            if (_buffer.Valid)
             {
-                _internalBuffer = ArrayPool<byte>.Shared.Rent(ReadBufferSize);
-                Debug.Assert(_internalOffset == 0);
-                Debug.Assert(_internalBufferCount == 0);
-            }
-            else if (_internalOffset > 0)
-            {
-                // We have buffered data at a non-zero offset.
+                // We may have buffered data at a non-zero offset.
                 // To maximize the buffer space available for the next read,
                 // copy the existing data down to the beginning of the buffer.
-                Buffer.BlockCopy(_internalBuffer, _internalOffset, _internalBuffer, 0, _internalBufferCount);
-                _internalOffset = 0;
+               _buffer.EnsureAvailableSpace(_buffer.Capacity - _buffer.ActiveLength);
```

Comment:
I don't think is same. Even if there is space in the buffer, whatever was left may be at the end ... limiting size of next IO. It would be interesting so how this impacts perf but if the chunk is small and copy fast the previous code would allow to fill the buffer completely.

Reply:
`EnsureAvailableSpace` moves the data to the beginning of the buffer if enough available space can be gained that way (see `ArrayBuffer.EnsureAvailableSpace`)

I haven't seen any significant change in perf benchmarks (changes were up to 2% in either direction)